{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas\n",
    "\n",
    "pandas.options.display.max_columns=999"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Load mutations data set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UNIQUEID</th>\n",
       "      <th>GENE</th>\n",
       "      <th>MUTATION</th>\n",
       "      <th>POSITION</th>\n",
       "      <th>AMINO_ACID_NUMBER</th>\n",
       "      <th>GENOME_INDEX</th>\n",
       "      <th>NUCLEOTIDE_NUMBER</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>IS_SNP</th>\n",
       "      <th>IS_INDEL</th>\n",
       "      <th>IN_CDS</th>\n",
       "      <th>IN_PROMOTER</th>\n",
       "      <th>IS_SYNONYMOUS</th>\n",
       "      <th>IS_NONSYNONYMOUS</th>\n",
       "      <th>IS_HET</th>\n",
       "      <th>IS_NULL</th>\n",
       "      <th>IS_FILTER_PASS</th>\n",
       "      <th>ELEMENT_TYPE</th>\n",
       "      <th>MUTATION_TYPE</th>\n",
       "      <th>INDEL_LENGTH</th>\n",
       "      <th>INDEL_1</th>\n",
       "      <th>INDEL_2</th>\n",
       "      <th>SITEID</th>\n",
       "      <th>NUMBER_NUCLEOTIDE_CHANGES</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>site.02.subj.0958.lab.22A197.iso.1</td>\n",
       "      <td>rpoB</td>\n",
       "      <td>P45S</td>\n",
       "      <td>45.0</td>\n",
       "      <td>45.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ccg</td>\n",
       "      <td>tcg</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>GENE</td>\n",
       "      <td>AAM</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>02</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>site.02.subj.0958.lab.22A197.iso.1</td>\n",
       "      <td>rpoB</td>\n",
       "      <td>S450L</td>\n",
       "      <td>450.0</td>\n",
       "      <td>450.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>tcg</td>\n",
       "      <td>ttg</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>GENE</td>\n",
       "      <td>AAM</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>02</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>site.02.subj.0958.lab.22A197.iso.1</td>\n",
       "      <td>rpoB</td>\n",
       "      <td>A1075A</td>\n",
       "      <td>1075.0</td>\n",
       "      <td>1075.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gct</td>\n",
       "      <td>gcc</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>GENE</td>\n",
       "      <td>AAM</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>02</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>site.02.subj.0958.lab.22A197.iso.1</td>\n",
       "      <td>rpoC</td>\n",
       "      <td>D271E</td>\n",
       "      <td>271.0</td>\n",
       "      <td>271.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gac</td>\n",
       "      <td>gag</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>GENE</td>\n",
       "      <td>AAM</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>02</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                             UNIQUEID  GENE MUTATION  POSITION  \\\n",
       "0  site.02.subj.0958.lab.22A197.iso.1  rpoB     P45S      45.0   \n",
       "1  site.02.subj.0958.lab.22A197.iso.1  rpoB    S450L     450.0   \n",
       "2  site.02.subj.0958.lab.22A197.iso.1  rpoB   A1075A    1075.0   \n",
       "3  site.02.subj.0958.lab.22A197.iso.1  rpoC    D271E     271.0   \n",
       "\n",
       "   AMINO_ACID_NUMBER  GENOME_INDEX  NUCLEOTIDE_NUMBER  REF  ALT  IS_SNP  \\\n",
       "0               45.0           NaN                NaN  ccg  tcg    True   \n",
       "1              450.0           NaN                NaN  tcg  ttg    True   \n",
       "2             1075.0           NaN                NaN  gct  gcc    True   \n",
       "3              271.0           NaN                NaN  gac  gag    True   \n",
       "\n",
       "   IS_INDEL  IN_CDS  IN_PROMOTER  IS_SYNONYMOUS  IS_NONSYNONYMOUS  IS_HET  \\\n",
       "0     False    True        False          False              True   False   \n",
       "1     False    True        False          False              True   False   \n",
       "2     False    True        False           True             False   False   \n",
       "3     False    True        False          False              True   False   \n",
       "\n",
       "   IS_NULL  IS_FILTER_PASS ELEMENT_TYPE MUTATION_TYPE  INDEL_LENGTH INDEL_1  \\\n",
       "0    False            True         GENE           AAM           NaN           \n",
       "1    False            True         GENE           AAM           NaN           \n",
       "2    False            True         GENE           AAM           NaN           \n",
       "3    False            True         GENE           AAM           NaN           \n",
       "\n",
       "  INDEL_2 SITEID  NUMBER_NUCLEOTIDE_CHANGES  \n",
       "0             02                          1  \n",
       "1             02                          1  \n",
       "2             02                          1  \n",
       "3             02                          1  "
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MUTATIONS = pandas.read_pickle('tables/MUTATIONS.pkl.gz')\n",
    "MUTATIONS.reset_index(inplace=True)\n",
    "MUTATIONS = MUTATIONS[(MUTATIONS.IS_FILTER_PASS) & (~MUTATIONS.IS_HET) & (~MUTATIONS.IS_NULL)]\n",
    "MUTATIONS[:4]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Load effects data set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 252,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "pandas.core.frame.DataFrame"
      ]
     },
     "execution_count": 252,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "EFFECTS = pandas.read_pickle('tables/EFFECTS.pkl.gz')\n",
    "EFFECTS.reset_index(inplace=True)\n",
    "EFFECTS[:20]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Find all samples with resistance mutations in EFFECTS and save UNIQUEID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3         site.02.subj.0958.lab.22A197.iso.1\n",
       "16        site.02.subj.0918.lab.22A153.iso.1\n",
       "45    site.02.subj.1033.lab.2013185075.iso.1\n",
       "51    site.02.subj.0739.lab.2013221518.iso.1\n",
       "89        site.02.subj.0104.lab.22A057.iso.1\n",
       "Name: UNIQUEID, dtype: object"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res_samples = EFFECTS[(EFFECTS.PREDICTION=='R') & (~EFFECTS.MUTATION.str[-1].isin(['O','X']))].UNIQUEID\n",
    "res_samples.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Filter out all mutations in samples that have resistance mutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mutation</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>P45S</td>\n",
       "      <td>rpoB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>S450L</td>\n",
       "      <td>rpoB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A1075A</td>\n",
       "      <td>rpoB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>D271E</td>\n",
       "      <td>rpoC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>D435G</td>\n",
       "      <td>rpoB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>517503</th>\n",
       "      <td>A172V</td>\n",
       "      <td>rpoC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>517504</th>\n",
       "      <td>R173R</td>\n",
       "      <td>rpoC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>517505</th>\n",
       "      <td>V1252L</td>\n",
       "      <td>rpoC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>517509</th>\n",
       "      <td>D435V</td>\n",
       "      <td>rpoB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>517510</th>\n",
       "      <td>A1075A</td>\n",
       "      <td>rpoB</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>68948 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       mutation  gene\n",
       "0          P45S  rpoB\n",
       "1         S450L  rpoB\n",
       "2        A1075A  rpoB\n",
       "3         D271E  rpoC\n",
       "8         D435G  rpoB\n",
       "...         ...   ...\n",
       "517503    A172V  rpoC\n",
       "517504    R173R  rpoC\n",
       "517505   V1252L  rpoC\n",
       "517509    D435V  rpoB\n",
       "517510   A1075A  rpoB\n",
       "\n",
       "[68948 rows x 2 columns]"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "co_occurr=pandas.DataFrame()\n",
    "co_occurr['mutation'] = MUTATIONS.loc[MUTATIONS['UNIQUEID'].isin(res_samples)].MUTATION\n",
    "co_occurr['gene']=MUTATIONS.loc[MUTATIONS['UNIQUEID'].isin(res_samples)].GENE\n",
    "co_occurr"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Remove mutations from list that are resistance mutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "mutation  gene\n",
       "A1075A    rpoB    12240\n",
       "A542A     rpoC     2880\n",
       "E1092D    rpoC     2439\n",
       "G594E     rpoC     1864\n",
       "G876G     rpoB     1458\n",
       "c-61t     rpoB     1402\n",
       "V483G     rpoC     1236\n",
       "D103D     rpoB     1041\n",
       "I491V     rpoC      666\n",
       "V483A     rpoC      599\n",
       "R173R     rpoC      595\n",
       "A172V     rpoC      589\n",
       "I491T     rpoC      457\n",
       "P1040R    rpoC      400\n",
       "F452S     rpoC      346\n",
       "C62C      rpoC      337\n",
       "E761D     rpoB      304\n",
       "P601L     rpoC      248\n",
       "E561E     rpoB      229\n",
       "L731P     rpoB      226\n",
       "dtype: int64"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res_mut = pandas.DataFrame()\n",
    "res_mut['mutation'] = EFFECTS[(EFFECTS.PREDICTION=='R') & (~EFFECTS.MUTATION.str[-1].isin(['O','X']))].MUTATION\n",
    "res_mut['gene'] = EFFECTS[(EFFECTS.PREDICTION=='R') & (~EFFECTS.MUTATION.str[-1].isin(['O','X']))].GENE\n",
    "co_occurr_only = co_occurr.loc[~co_occurr['mutation'].isin(res_mut['mutation'])]\n",
    "co_occurr_only.value_counts()[:20]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Either** look only at co-occuring mutations that do not occur without resistance mutations:\n",
    "\n",
    "- first determine how often mutations that co-occurr are found in general"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "mutation  gene\n",
       "A1075A    rpoB    37400\n",
       "G594E     rpoC    10610\n",
       "A542A     rpoC    10544\n",
       "G876G     rpoB     9236\n",
       "c-61t     rpoB     8752\n",
       "                  ...  \n",
       "S401P     rpoC        1\n",
       "F438L     sigA        1\n",
       "F424I     rpoB        1\n",
       "F424C     rpoB        1\n",
       "E830Q     rpoC        1\n",
       "Length: 3149, dtype: int64"
      ]
     },
     "execution_count": 151,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gen_occurr=pandas.DataFrame()\n",
    "gen_occurr['mutation']=MUTATIONS.loc[MUTATIONS['MUTATION'].isin(co_occurr_only.mutation)].MUTATION\n",
    "gen_occurr['gene']=MUTATIONS.loc[MUTATIONS['MUTATION'].isin(co_occurr_only.mutation)].GENE\n",
    "gen_occurr.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 263,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mutation</th>\n",
       "      <th>gene</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A1075A</td>\n",
       "      <td>rpoB</td>\n",
       "      <td>37400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>G594E</td>\n",
       "      <td>rpoC</td>\n",
       "      <td>10610</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A542A</td>\n",
       "      <td>rpoC</td>\n",
       "      <td>10544</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>G876G</td>\n",
       "      <td>rpoB</td>\n",
       "      <td>9236</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>c-61t</td>\n",
       "      <td>rpoB</td>\n",
       "      <td>8752</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3144</th>\n",
       "      <td>S401P</td>\n",
       "      <td>rpoC</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3145</th>\n",
       "      <td>F438L</td>\n",
       "      <td>sigA</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3146</th>\n",
       "      <td>F424I</td>\n",
       "      <td>rpoB</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3147</th>\n",
       "      <td>F424C</td>\n",
       "      <td>rpoB</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3148</th>\n",
       "      <td>E830Q</td>\n",
       "      <td>rpoC</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3149 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     mutation  gene  count\n",
       "0      A1075A  rpoB  37400\n",
       "1       G594E  rpoC  10610\n",
       "2       A542A  rpoC  10544\n",
       "3       G876G  rpoB   9236\n",
       "4       c-61t  rpoB   8752\n",
       "...       ...   ...    ...\n",
       "3144    S401P  rpoC      1\n",
       "3145    F438L  sigA      1\n",
       "3146    F424I  rpoB      1\n",
       "3147    F424C  rpoB      1\n",
       "3148    E830Q  rpoC      1\n",
       "\n",
       "[3149 rows x 3 columns]"
      ]
     },
     "execution_count": 263,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_gen=gen_occurr.value_counts().reset_index(name='count')\n",
    "df_res=co_occurr_only.value_counts().reset_index(name='count')\n",
    "df_gen\n",
    "#gen_occurr.value_counts()-co_occurr_only.value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- then identify if difference of the general occurrence and occurrence with resistance mutations is zero, if yes, mutations only occurr with resistance mutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 271,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mutation</th>\n",
       "      <th>gene</th>\n",
       "      <th>count</th>\n",
       "      <th>diff</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2619</th>\n",
       "      <td>-11_indel</td>\n",
       "      <td>rpoZ</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2617</th>\n",
       "      <td>-3_indel</td>\n",
       "      <td>rpoA</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2632</th>\n",
       "      <td>1023_indel</td>\n",
       "      <td>rpoA</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2624</th>\n",
       "      <td>1286_indel</td>\n",
       "      <td>rpoB</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2744</th>\n",
       "      <td>1361_indel</td>\n",
       "      <td>rpoB</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2637</th>\n",
       "      <td>g-60a</td>\n",
       "      <td>rpoZ</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2640</th>\n",
       "      <td>g-7a</td>\n",
       "      <td>sigA</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2611</th>\n",
       "      <td>g-92t</td>\n",
       "      <td>rpoB</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2626</th>\n",
       "      <td>t-45g</td>\n",
       "      <td>rpoB</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2638</th>\n",
       "      <td>t-88c</td>\n",
       "      <td>sigA</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>455 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        mutation  gene  count  diff\n",
       "2619   -11_indel  rpoZ      1   0.0\n",
       "2617    -3_indel  rpoA      1   0.0\n",
       "2632  1023_indel  rpoA      1   0.0\n",
       "2624  1286_indel  rpoB      1   0.0\n",
       "2744  1361_indel  rpoB      1   0.0\n",
       "...          ...   ...    ...   ...\n",
       "2637       g-60a  rpoZ      1   0.0\n",
       "2640        g-7a  sigA      1   0.0\n",
       "2611       g-92t  rpoB      1   0.0\n",
       "2626       t-45g  rpoB      1   0.0\n",
       "2638       t-88c  sigA      1   0.0\n",
       "\n",
       "[455 rows x 4 columns]"
      ]
     },
     "execution_count": 271,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#df_gen.set_index('mutation')\n",
    "#df_res.set_index('mutation')\n",
    "df_gen.sort_values('mutation',inplace=True)\n",
    "df_res.sort_values('mutation',inplace=True)\n",
    "\n",
    "df_gen['diff']=df_gen['count']-df_res['count']\n",
    "df_gen[(df_gen['diff']==0)]\n",
    "#df_gen[(df_gen['diff']==0)&(df_gen['count']>1)]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Or** determine if co-occurrence is significant compared to general occurrence in mutations dataset:\n",
    "\n",
    "- find number of samples in mutations set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 159,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "64722"
      ]
     },
     "execution_count": 159,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(MUTATIONS.UNIQUEID.value_counts()) "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- find number of samples with resistance mutations in effects dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 160,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "21046"
      ]
     },
     "execution_count": 160,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(EFFECTS[(EFFECTS.PREDICTION=='R')].UNIQUEID.value_counts())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- determine relative occurrence of mutations in co-occurr list in general mutations set and in resistant samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 161,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0           P45S\n",
       "2         A1075A\n",
       "3          D271E\n",
       "9          I491L\n",
       "10        A1075A\n",
       "           ...  \n",
       "517502      Y61Y\n",
       "517503     A172V\n",
       "517504     R173R\n",
       "517505    V1252L\n",
       "517510    A1075A\n",
       "Name: mutation, Length: 48118, dtype: object"
      ]
     },
     "execution_count": 161,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "co_occurr_only.mutation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 162,
   "metadata": {},
   "outputs": [],
   "source": [
    "#gen_occurr=pandas.DataFrame()\n",
    "#gen_occurr['mutation']=MUTATIONS.loc[MUTATIONS['MUTATION'].isin(co_occurr_only.mutation)].MUTATION\n",
    "#gen_occurr['gene']=MUTATIONS.loc[MUTATIONS['MUTATION'].isin(co_occurr_only.mutation)].GENE\n",
    "#gen_occurr.value_counts()[:20]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 163,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mutation  gene\n",
      "A1075A    rpoB    0.581583\n",
      "A542A     rpoC    0.136843\n",
      "E1092D    rpoC    0.115889\n",
      "G594E     rpoC    0.088568\n",
      "G876G     rpoB    0.069277\n",
      "c-61t     rpoB    0.066616\n",
      "V483G     rpoC    0.058728\n",
      "D103D     rpoB    0.049463\n",
      "I491V     rpoC    0.031645\n",
      "V483A     rpoC    0.028461\n",
      "R173R     rpoC    0.028271\n",
      "A172V     rpoC    0.027986\n",
      "I491T     rpoC    0.021714\n",
      "P1040R    rpoC    0.019006\n",
      "F452S     rpoC    0.016440\n",
      "C62C      rpoC    0.016013\n",
      "E761D     rpoB    0.014445\n",
      "P601L     rpoC    0.011784\n",
      "E561E     rpoB    0.010881\n",
      "L731P     rpoB    0.010738\n",
      "dtype: float64\n",
      "mutation  gene\n",
      "A1075A    rpoB    0.577856\n",
      "G594E     rpoC    0.163932\n",
      "A542A     rpoC    0.162912\n",
      "G876G     rpoB    0.142703\n",
      "c-61t     rpoB    0.135224\n",
      "D103D     rpoB    0.104400\n",
      "R173R     rpoC    0.103643\n",
      "A172V     rpoC    0.101496\n",
      "E1092D    rpoC    0.061818\n",
      "P601L     rpoC    0.060180\n",
      "P54P      rpoC    0.020828\n",
      "V483G     rpoC    0.019205\n",
      "V228V     rpoB    0.016795\n",
      "A621T     rpoC    0.015142\n",
      "E319K     rpoA    0.014400\n",
      "I491V     rpoC    0.010568\n",
      "V298V     rpoC    0.010136\n",
      "V483A     rpoC    0.009564\n",
      "P906A     rpoC    0.009162\n",
      "E561E     rpoB    0.008822\n",
      "dtype: float64\n"
     ]
    }
   ],
   "source": [
    "rel_occurr_res=co_occurr_only.value_counts().div(21046)\n",
    "rel_occurr_gen=gen_occurr.value_counts().div(64722)\n",
    "print(rel_occurr_res[:20])\n",
    "print(rel_occurr_gen[:20])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 164,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>resistance</th>\n",
       "      <th>general</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mutation</th>\n",
       "      <th>gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1075A</th>\n",
       "      <th>rpoB</th>\n",
       "      <td>0.581583</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A542A</th>\n",
       "      <th>rpoC</th>\n",
       "      <td>0.136843</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E1092D</th>\n",
       "      <th>rpoC</th>\n",
       "      <td>0.115889</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G594E</th>\n",
       "      <th>rpoC</th>\n",
       "      <td>0.088568</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G876G</th>\n",
       "      <th>rpoB</th>\n",
       "      <td>0.069277</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R665W</th>\n",
       "      <th>rpoB</th>\n",
       "      <td>0.000048</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>L181I</th>\n",
       "      <th>rpoC</th>\n",
       "      <td>0.000048</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>L194L</th>\n",
       "      <th>rpoA</th>\n",
       "      <td>0.000048</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F1175C</th>\n",
       "      <th>rpoC</th>\n",
       "      <td>0.000048</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G519G</th>\n",
       "      <th>rpoC</th>\n",
       "      <td>0.000048</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3063 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               resistance  general\n",
       "mutation gene                     \n",
       "A1075A   rpoB    0.581583      NaN\n",
       "A542A    rpoC    0.136843      NaN\n",
       "E1092D   rpoC    0.115889      NaN\n",
       "G594E    rpoC    0.088568      NaN\n",
       "G876G    rpoB    0.069277      NaN\n",
       "...                   ...      ...\n",
       "R665W    rpoB    0.000048      NaN\n",
       "L181I    rpoC    0.000048      NaN\n",
       "L194L    rpoA    0.000048      NaN\n",
       "F1175C   rpoC    0.000048      NaN\n",
       "G519G    rpoC    0.000048      NaN\n",
       "\n",
       "[3063 rows x 2 columns]"
      ]
     },
     "execution_count": 164,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "concat_rel_occurr=pandas.DataFrame()\n",
    "concat_rel_occurr['resistance']=rel_occurr_res\n",
    "concat_rel_occurr['general']=rel_occur_gen\n",
    "concat_rel_occurr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
