{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "40a64925-297b-4a45-987d-6c2651d3901e",
   "metadata": {},
   "source": [
    "# tb-rnap-compensation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "53d58ddd-781f-4190-9462-d55f4989ea1d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas\n",
    "\n",
    "pandas.options.display.max_columns=999"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4bfb6bd8-82a6-4641-b505-35254788c884",
   "metadata": {},
   "source": [
    "Let's load in the `MUTATIONS` table and have a look. Importantly, this table also records `NULL`s (where there are no reads at an amino acid so we have no evidence of what is there) and `FILTER_FAIL`s (where is some evidence but not enough to be statistically significant). These need excluding."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "7c7e7bb9-d480-4add-ad7e-d2b76c24d139",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UNIQUEID</th>\n",
       "      <th>GENE</th>\n",
       "      <th>MUTATION</th>\n",
       "      <th>POSITION</th>\n",
       "      <th>AMINO_ACID_NUMBER</th>\n",
       "      <th>GENOME_INDEX</th>\n",
       "      <th>NUCLEOTIDE_NUMBER</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>IS_SNP</th>\n",
       "      <th>IS_INDEL</th>\n",
       "      <th>IN_CDS</th>\n",
       "      <th>IN_PROMOTER</th>\n",
       "      <th>IS_SYNONYMOUS</th>\n",
       "      <th>IS_NONSYNONYMOUS</th>\n",
       "      <th>IS_HET</th>\n",
       "      <th>IS_NULL</th>\n",
       "      <th>IS_FILTER_PASS</th>\n",
       "      <th>ELEMENT_TYPE</th>\n",
       "      <th>MUTATION_TYPE</th>\n",
       "      <th>INDEL_LENGTH</th>\n",
       "      <th>INDEL_1</th>\n",
       "      <th>INDEL_2</th>\n",
       "      <th>SITEID</th>\n",
       "      <th>NUMBER_NUCLEOTIDE_CHANGES</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>site.02.subj.0958.lab.22A197.iso.1</td>\n",
       "      <td>rpoB</td>\n",
       "      <td>P45S</td>\n",
       "      <td>45.0</td>\n",
       "      <td>45.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ccg</td>\n",
       "      <td>tcg</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>GENE</td>\n",
       "      <td>AAM</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>02</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>site.02.subj.0958.lab.22A197.iso.1</td>\n",
       "      <td>rpoB</td>\n",
       "      <td>S450L</td>\n",
       "      <td>450.0</td>\n",
       "      <td>450.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>tcg</td>\n",
       "      <td>ttg</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>GENE</td>\n",
       "      <td>AAM</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>02</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>site.02.subj.0958.lab.22A197.iso.1</td>\n",
       "      <td>rpoB</td>\n",
       "      <td>A1075A</td>\n",
       "      <td>1075.0</td>\n",
       "      <td>1075.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gct</td>\n",
       "      <td>gcc</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>GENE</td>\n",
       "      <td>AAM</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>02</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>site.02.subj.0958.lab.22A197.iso.1</td>\n",
       "      <td>rpoC</td>\n",
       "      <td>D271E</td>\n",
       "      <td>271.0</td>\n",
       "      <td>271.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gac</td>\n",
       "      <td>gag</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>GENE</td>\n",
       "      <td>AAM</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>02</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                             UNIQUEID  GENE MUTATION  POSITION  \\\n",
       "0  site.02.subj.0958.lab.22A197.iso.1  rpoB     P45S      45.0   \n",
       "1  site.02.subj.0958.lab.22A197.iso.1  rpoB    S450L     450.0   \n",
       "2  site.02.subj.0958.lab.22A197.iso.1  rpoB   A1075A    1075.0   \n",
       "3  site.02.subj.0958.lab.22A197.iso.1  rpoC    D271E     271.0   \n",
       "\n",
       "   AMINO_ACID_NUMBER  GENOME_INDEX  NUCLEOTIDE_NUMBER  REF  ALT  IS_SNP  \\\n",
       "0               45.0           NaN                NaN  ccg  tcg    True   \n",
       "1              450.0           NaN                NaN  tcg  ttg    True   \n",
       "2             1075.0           NaN                NaN  gct  gcc    True   \n",
       "3              271.0           NaN                NaN  gac  gag    True   \n",
       "\n",
       "   IS_INDEL  IN_CDS  IN_PROMOTER  IS_SYNONYMOUS  IS_NONSYNONYMOUS  IS_HET  \\\n",
       "0     False    True        False          False              True   False   \n",
       "1     False    True        False          False              True   False   \n",
       "2     False    True        False           True             False   False   \n",
       "3     False    True        False          False              True   False   \n",
       "\n",
       "   IS_NULL  IS_FILTER_PASS ELEMENT_TYPE MUTATION_TYPE  INDEL_LENGTH INDEL_1  \\\n",
       "0    False            True         GENE           AAM           NaN           \n",
       "1    False            True         GENE           AAM           NaN           \n",
       "2    False            True         GENE           AAM           NaN           \n",
       "3    False            True         GENE           AAM           NaN           \n",
       "\n",
       "  INDEL_2 SITEID  NUMBER_NUCLEOTIDE_CHANGES  \n",
       "0             02                          1  \n",
       "1             02                          1  \n",
       "2             02                          1  \n",
       "3             02                          1  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MUTATIONS = pandas.read_pickle('tables/MUTATIONS.pkl.gz')\n",
    "MUTATIONS.reset_index(inplace=True)\n",
    "MUTATIONS = MUTATIONS[(MUTATIONS.IS_FILTER_PASS) & (~MUTATIONS.IS_HET) & (~MUTATIONS.IS_NULL)]\n",
    "MUTATIONS[:4]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "04c00f04-bb07-429d-890c-e7533ae3f9d5",
   "metadata": {},
   "source": [
    "To get a quick feel for the amount of reversion that may be happening, let's do a quick cross-tab"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "b011cf5b-0702-4008-b7e9-3e554843fc4a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>NUMBER_NUCLEOTIDE_CHANGES</th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GENE</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>rpoA</th>\n",
       "      <td>219</td>\n",
       "      <td>15384</td>\n",
       "      <td>429</td>\n",
       "      <td>41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rpoB</th>\n",
       "      <td>9491</td>\n",
       "      <td>115155</td>\n",
       "      <td>3053</td>\n",
       "      <td>911</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rpoC</th>\n",
       "      <td>1175</td>\n",
       "      <td>101868</td>\n",
       "      <td>3631</td>\n",
       "      <td>1108</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rpoZ</th>\n",
       "      <td>102</td>\n",
       "      <td>1593</td>\n",
       "      <td>32</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sigA</th>\n",
       "      <td>343</td>\n",
       "      <td>13279</td>\n",
       "      <td>395</td>\n",
       "      <td>79</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "NUMBER_NUCLEOTIDE_CHANGES     0       1     2     3\n",
       "GENE                                               \n",
       "rpoA                        219   15384   429    41\n",
       "rpoB                       9491  115155  3053   911\n",
       "rpoC                       1175  101868  3631  1108\n",
       "rpoZ                        102    1593    32     1\n",
       "sigA                        343   13279   395    79"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pandas.crosstab(MUTATIONS.GENE, MUTATIONS.NUMBER_NUCLEOTIDE_CHANGES)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "08e68749-38c5-448a-89bc-7d67ae1330f8",
   "metadata": {},
   "source": [
    "Sure enough there are a good number of codons in `rpoB` and `rpoC` which have two or three bases different to the reference genome.\n",
    "\n",
    "Let's look at those were two bases are different as they fit our hypothesis (harder to explain three!)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "801d4560-1884-448f-8460-de2f5cd2a92f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UNIQUEID</th>\n",
       "      <th>GENE</th>\n",
       "      <th>MUTATION</th>\n",
       "      <th>POSITION</th>\n",
       "      <th>AMINO_ACID_NUMBER</th>\n",
       "      <th>GENOME_INDEX</th>\n",
       "      <th>NUCLEOTIDE_NUMBER</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>IS_SNP</th>\n",
       "      <th>IS_INDEL</th>\n",
       "      <th>IN_CDS</th>\n",
       "      <th>IN_PROMOTER</th>\n",
       "      <th>IS_SYNONYMOUS</th>\n",
       "      <th>IS_NONSYNONYMOUS</th>\n",
       "      <th>IS_HET</th>\n",
       "      <th>IS_NULL</th>\n",
       "      <th>IS_FILTER_PASS</th>\n",
       "      <th>ELEMENT_TYPE</th>\n",
       "      <th>MUTATION_TYPE</th>\n",
       "      <th>INDEL_LENGTH</th>\n",
       "      <th>INDEL_1</th>\n",
       "      <th>INDEL_2</th>\n",
       "      <th>SITEID</th>\n",
       "      <th>NUMBER_NUCLEOTIDE_CHANGES</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>817</th>\n",
       "      <td>site.02.subj.0926.lab.22A161.iso.1</td>\n",
       "      <td>rpoB</td>\n",
       "      <td>S450F</td>\n",
       "      <td>450.0</td>\n",
       "      <td>450.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>tcg</td>\n",
       "      <td>ttc</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>GENE</td>\n",
       "      <td>AAM</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>02</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>859</th>\n",
       "      <td>site.02.subj.0893.lab.22A127.iso.1</td>\n",
       "      <td>rpoB</td>\n",
       "      <td>D435F</td>\n",
       "      <td>435.0</td>\n",
       "      <td>435.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gac</td>\n",
       "      <td>ttc</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>GENE</td>\n",
       "      <td>AAM</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>02</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1022</th>\n",
       "      <td>site.02.subj.0197.lab.2013221241.iso.1</td>\n",
       "      <td>sigA</td>\n",
       "      <td>A55S</td>\n",
       "      <td>55.0</td>\n",
       "      <td>55.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gcc</td>\n",
       "      <td>tcg</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>GENE</td>\n",
       "      <td>AAM</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>02</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1506</th>\n",
       "      <td>site.02.subj.0074.lab.22A026.iso.1</td>\n",
       "      <td>rpoB</td>\n",
       "      <td>H445C</td>\n",
       "      <td>445.0</td>\n",
       "      <td>445.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>cac</td>\n",
       "      <td>tgc</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>GENE</td>\n",
       "      <td>AAM</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>02</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2244</th>\n",
       "      <td>site.05.subj.LR-2335.lab.FN-01418-18.iso.1</td>\n",
       "      <td>rpoB</td>\n",
       "      <td>S450M</td>\n",
       "      <td>450.0</td>\n",
       "      <td>450.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>tcg</td>\n",
       "      <td>atg</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>GENE</td>\n",
       "      <td>AAM</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>05</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                        UNIQUEID  GENE MUTATION  POSITION  \\\n",
       "817           site.02.subj.0926.lab.22A161.iso.1  rpoB    S450F     450.0   \n",
       "859           site.02.subj.0893.lab.22A127.iso.1  rpoB    D435F     435.0   \n",
       "1022      site.02.subj.0197.lab.2013221241.iso.1  sigA     A55S      55.0   \n",
       "1506          site.02.subj.0074.lab.22A026.iso.1  rpoB    H445C     445.0   \n",
       "2244  site.05.subj.LR-2335.lab.FN-01418-18.iso.1  rpoB    S450M     450.0   \n",
       "\n",
       "      AMINO_ACID_NUMBER  GENOME_INDEX  NUCLEOTIDE_NUMBER  REF  ALT  IS_SNP  \\\n",
       "817               450.0           NaN                NaN  tcg  ttc    True   \n",
       "859               435.0           NaN                NaN  gac  ttc    True   \n",
       "1022               55.0           NaN                NaN  gcc  tcg    True   \n",
       "1506              445.0           NaN                NaN  cac  tgc    True   \n",
       "2244              450.0           NaN                NaN  tcg  atg    True   \n",
       "\n",
       "      IS_INDEL  IN_CDS  IN_PROMOTER  IS_SYNONYMOUS  IS_NONSYNONYMOUS  IS_HET  \\\n",
       "817      False    True        False          False              True   False   \n",
       "859      False    True        False          False              True   False   \n",
       "1022     False    True        False          False              True   False   \n",
       "1506     False    True        False          False              True   False   \n",
       "2244     False    True        False          False              True   False   \n",
       "\n",
       "      IS_NULL  IS_FILTER_PASS ELEMENT_TYPE MUTATION_TYPE  INDEL_LENGTH  \\\n",
       "817     False            True         GENE           AAM           NaN   \n",
       "859     False            True         GENE           AAM           NaN   \n",
       "1022    False            True         GENE           AAM           NaN   \n",
       "1506    False            True         GENE           AAM           NaN   \n",
       "2244    False            True         GENE           AAM           NaN   \n",
       "\n",
       "     INDEL_1 INDEL_2 SITEID  NUMBER_NUCLEOTIDE_CHANGES  \n",
       "817                      02                          2  \n",
       "859                      02                          2  \n",
       "1022                     02                          2  \n",
       "1506                     02                          2  \n",
       "2244                     05                          2  "
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MUTATIONS[MUTATIONS.NUMBER_NUCLEOTIDE_CHANGES == 2][:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "53e6f6ee-1e95-4053-8fa1-d98b5a4d27a7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "450.0    232\n",
       "445.0    160\n",
       "656.0    136\n",
       "640.0    135\n",
       "662.0    129\n",
       "243.0    102\n",
       "641.0     99\n",
       "549.0     97\n",
       "435.0     85\n",
       "681.0     78\n",
       "670.0     76\n",
       "545.0     76\n",
       "443.0     59\n",
       "544.0     56\n",
       "431.0     52\n",
       "Name: AMINO_ACID_NUMBER, dtype: int64"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "a = MUTATIONS[(MUTATIONS.GENE == 'rpoB') & (MUTATIONS.NUMBER_NUCLEOTIDE_CHANGES == 2)].AMINO_ACID_NUMBER.value_counts()\n",
    "a[a>50]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "541ef7a6-bec7-41d4-ad68-d0da188067e3",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
